kv/kvserver: TestStoreRangeMergeWithData failed

kv/kvserver.TestStoreRangeMergeWithData [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12394618?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12394618?buildTab=artifacts#/) on master @ [aa85c69776da942b4d539891725b780ba995401e](https://github.com/cockroachdb/cockroach/commits/aa85c69776da942b4d539891725b780ba995401e):

Fatal error:

```
panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/MuxRangeFeed. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/MuxRangeFeed. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 507456 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc011f3f1d0?, {0xd3af058, 0xc00ac0f0b0})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x74
panic({0x944f060?, 0xc00cd11630?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc007055c00)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x23d
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc007055c00)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442 +0x27
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0xd3af058, 0xc007870f30})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x45
github.com/cockroachdb/cockroach/pkg/util/log.HasSpanOrEvent(...)
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:274
github.com/cockroachdb/cockroach/pkg/storage.pebbleLogger.IsTracingEnabled(...)
	github.com/cockroachdb/cockroach/pkg/storage/pebble.go:743
github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0xc000f01880, {0xd3af058, 0xc007870f30}, {0xc01de48478?, 0x1b17cd5?}, 0x0, {0xd3b3ea8, 0xc008c5ac28}, 0xc01206f3a8, 0x0)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader.go:569 +0x5c5
github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).loadBlock(0xc008c5a900, 0xde?)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:398 +0x85b
github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).skipForward(0xc008c5a900)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:1246 +0x6b
github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).Next(0xc008c5a900)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:1137 +0x2ba
github.com/cockroachdb/pebble.(*levelIter).Next(0xc01206fcc0)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:911 +0x3a4
github.com/cockroachdb/pebble.(*mergingIter).nextEntry(0xc01206f5d8, 0xc01206f768, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:620 +0x592
github.com/cockroachdb/pebble.(*mergingIter).Next(0xc01206f5d8)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1254 +0xcd
github.com/cockroachdb/pebble.(*lazyCombinedIter).Next(0xc01206f440)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/range_keys.go:619 +0x7b
github.com/cockroachdb/pebble.(*Iterator).internalNextPrefix(0xc01206f000, 0x6)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1797 +0x103
github.com/cockroachdb/pebble.(*Iterator).nextPrefix(0xc01206f000)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1721 +0xd18
github.com/cockroachdb/pebble.(*Iterator).NextPrefix(0xc01206f000)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1687 +0x26d
github.com/cockroachdb/cockroach/pkg/storage/pebbleiter.(*assertionIter).NextPrefix(0xc012574280)
	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/pebbleiter/gen-crdb_test_on.go:201 +0x85
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).NextKey(0xc00608dd40)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:514 +0xd2
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).NextKey(0xc010b89700)
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:922 +0x4d6
github.com/cockroachdb/cockroach/pkg/storage.(*unsafeMVCCIterator).NextKey(0xc0131685a0)
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1524 +0x45
github.com/cockroachdb/cockroach/pkg/storage.(*MVCCIncrementalIterator).advance(0xc00393a1a0, 0x1)
	github.com/cockroachdb/cockroach/pkg/storage/mvcc_incremental_iterator.go:594 +0x67a
github.com/cockroachdb/cockroach/pkg/storage.(*MVCCIncrementalIterator).SeekGE(0xc00393a1a0, {{0xc019bb81d8, 0x1, 0x8}, {0x0, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/mvcc_incremental_iterator.go:272 +0x3dd
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*CatchUpIterator).CatchUpScan(0xc012e80000, {0xd3af020, 0xc00b084af0}, 0xc00f359be0, 0x1)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/catchup_scan.go:165 +0x23b
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).maybeRunCatchUpScan(0xc011f3f1d0, {0xd3af020, 0xc00b084af0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:400 +0x1ca
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).outputLoop(0xc011f3f1d0, {0xd3af020, 0xc00b084af0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:317 +0x65
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).runOutputLoop(0xc011f3f1d0, {0xd3af058, 0xc00ac0f0b0}, 0xc004ba4e10?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:361 +0x129
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*ScheduledProcessor).Register.func1.1({0xd3af058, 0xc00ac0f0b0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduled_processor.go:335 +0x68
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x1eb
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 144430
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x5f9
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | 	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader.go:569
*   | github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).loadBlock
*   | 	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:398
*   | github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).skipForward
*   | 	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:1246
*   | github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).Next
*   | 	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:1137
*   | github.com/cockroachdb/pebble.(*levelIter).Next
*   | 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:911
*   | github.com/cockroachdb/pebble.(*mergingIter).nextEntry
*   | 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:620
*   | github.com/cockroachdb/pebble.(*mergingIter).Next
*   | 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1254
*   | github.com/cockroachdb/pebble.(*lazyCombinedIter).Next
*   | 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/range_keys.go:619
*   | github.com/cockroachdb/pebble.(*Iterator).internalNextPrefix
*   | 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1797
*   | github.com/cockroachdb/pebble.(*Iterator).nextPrefix
*   | 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1721
*   | github.com/cockroachdb/pebble.(*Iterator).NextPrefix
*   | 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1687
*   | github.com/cockroachdb/cockroach/pkg/storage/pebbleiter.(*assertionIter).NextPrefix
*   | 	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/pebbleiter/gen-crdb_test_on.go:201
*   | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).NextKey
*   | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:514
*   | github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).NextKey
*   | 	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:922
*   | github.com/cockroachdb/cockroach/pkg/storage.(*unsafeMVCCIterator).NextKey
*   | 	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1524
*   | github.com/cockroachdb/cockroach/pkg/storage.(*MVCCIncrementalIterator).advance
*   | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc_incremental_iterator.go:594
*   | github.com/cockroachdb/cockroach/pkg/storage.(*MVCCIncrementalIterator).SeekGE
*   | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc_incremental_iterator.go:272
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*CatchUpIterator).CatchUpScan
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/catchup_scan.go:165
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).maybeRunCatchUpScan
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:400
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).outputLoop
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:317
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).runOutputLoop
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:361
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*ScheduledProcessor).Register.func1.1
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduled_processor.go:335
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1650
* Wraps: (4) use of Span after Finish. Span: /cockroach.roachpb.Internal/MuxRangeFeed. Finish previously called at: <stack not captured. Set debugUseAfterFinish>
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeWithData.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-32828

--- a/pkg/kv/kvserver/rangefeed/catchup_scan.go
+++ b/pkg/kv/kvserver/rangefeed/catchup_scan.go
@@ -135,6 +135,9 @@ type outputEventFn func(e *kvpb.RangeFeedEvent) error
 // For example, with MVCC range tombstones [a-f)@5 and [a-f)@3 overlapping point
 // keys a@6, a@4, and b@2, the emitted order is [a-f)@3,[a-f)@5,a@4,a@6,b@2 because
 // the start key "a" is ordered before all of the timestamped point keys.
+//
+// TODO(sumeer): ctx is not used for SeekGE and Next. Fix by adding a method
+// to SimpleMVCCIterator to replace the context.
 func (i *CatchUpIterator) CatchUpScan(
 	ctx context.Context, outputFn outputEventFn, withDiff bool,
 ) error {

--- a/pkg/kv/kvserver/replica_rangefeed.go
+++ b/pkg/kv/kvserver/replica_rangefeed.go
@@ -273,8 +273,10 @@ func (r *Replica) RangeFeed(
 	// Register the stream with a catch-up iterator.
 	var catchUpIter *rangefeed.CatchUpIterator
 	if usingCatchUpIter {
+		// Pass context.Background() since the context where the iter will be used
+		// is different.
 		catchUpIter, err = rangefeed.NewCatchUpIterator(
-			ctx, r.store.TODOEngine(), rSpan.AsRawSpanWithNoLocals(),
+			context.Background(), r.store.TODOEngine(), rSpan.AsRawSpanWithNoLocals(),
 			args.Timestamp, iterSemRelease, pacer)
 		if err != nil {
 			r.raftMu.Unlock()

--- a/pkg/storage/pebble.go
+++ b/pkg/storage/pebble.go
@@ -2635,7 +2635,7 @@ func (p *pebbleReadOnly) NewMVCCIterator(
 	}
 
 	if iter.iter != nil {
-		iter.setOptions(opts, p.durability)
+		iter.setOptions(ctx, opts, p.durability)
 	} else {
 		if err := iter.initReuseOrCreate(
 			ctx, p.parent.db, p.iter, p.iterUsed, opts, p.durability, p.parent); err != nil {
@@ -2673,7 +2673,7 @@ func (p *pebbleReadOnly) NewEngineIterator(
 	}
 
 	if iter.iter != nil {
-		iter.setOptions(opts, p.durability)
+		iter.setOptions(ctx, opts, p.durability)
 	} else {
 		err := iter.initReuseOrCreate(
 			ctx, p.parent.db, p.iter, p.iterUsed, opts, p.durability, p.parent)

--- a/pkg/storage/pebble_batch.go
+++ b/pkg/storage/pebble_batch.go
@@ -208,7 +208,7 @@ func (p *pebbleBatch) NewMVCCIterator(
 	}
 
 	if iter.iter != nil {
-		iter.setOptions(opts, StandardDurability)
+		iter.setOptions(ctx, opts, StandardDurability)
 	} else {
 		if err := iter.initReuseOrCreate(
 			ctx, handle, p.iter, p.iterUsed, opts, StandardDurability, p.parent); err != nil {
@@ -249,7 +249,7 @@ func (p *pebbleBatch) NewEngineIterator(
 	}
 
 	if iter.iter != nil {
-		iter.setOptions(opts, StandardDurability)
+		iter.setOptions(ctx, opts, StandardDurability)
 	} else {
 		if err := iter.initReuseOrCreate(
 			ctx, handle, p.iter, p.iterUsed, opts, StandardDurability, p.parent); err != nil {

--- a/pkg/storage/pebble_iterator.go
+++ b/pkg/storage/pebble_iterator.go
@@ -104,9 +104,8 @@ func newPebbleIterator(
 ) (*pebbleIterator, error) {
 	p := pebbleIterPool.Get().(*pebbleIterator)
 	p.reusable = false // defensive
-	p.init(nil, opts, durability, parent)
-	// TODO(sumeer): fix after bumping to latest Pebble.
-	iter, err := handle.NewIter(&p.options)
+	p.init(ctx, nil, opts, durability, parent)
+	iter, err := handle.NewIterWithContext(ctx, &p.options)
 	if err != nil {
 		return nil, err
 	}
@@ -122,7 +121,7 @@ func newPebbleIteratorByCloning(
 	var err error
 	p := pebbleIterPool.Get().(*pebbleIterator)
 	p.reusable = false // defensive
-	p.init(nil, opts, durability, cloneCtx.engine)
+	p.init(ctx, nil, opts, durability, cloneCtx.engine)
 	p.iter, err = cloneCtx.rawIter.CloneWithContext(ctx, pebble.CloneOptions{
 		IterOptions:      &p.options,
 		RefreshBatchView: true,
@@ -140,7 +139,7 @@ func newPebbleSSTIterator(
 ) (*pebbleIterator, error) {
 	p := pebbleIterPool.Get().(*pebbleIterator)
 	p.reusable = false // defensive
-	p.init(nil, opts, StandardDurability, nil)
+	p.init(context.Background(), nil, opts, StandardDurability, nil)
 
 	var externalIterOpts []pebble.ExternalIterOption
 	if forwardOnly {
@@ -161,6 +160,7 @@ func newPebbleSSTIterator(
 // reconfiguring the given iter. It is valid to pass a nil iter and then create
 // p.iter using p.options, to avoid redundant reconfiguration via SetOptions().
 func (p *pebbleIterator) init(
+	ctx context.Context,
 	iter pebbleiter.Iterator,
 	opts IterOptions,
 	durability DurabilityRequirement,
@@ -175,7 +175,7 @@ func (p *pebbleIterator) init(
 		parent:             statsReporter,
 		reusable:           p.reusable,
 	}
-	p.setOptions(opts, durability)
+	p.setOptions(ctx, opts, durability)
 	p.inuse = true // after setOptions(), so panic won't cause reader to panic too
 }
 
@@ -195,14 +195,14 @@ func (p *pebbleIterator) initReuseOrCreate(
 	statsReporter *Pebble,
 ) error {
 	if iter != nil && !clone {
-		p.init(iter, opts, durability, statsReporter)
+		p.init(ctx, iter, opts, durability, statsReporter)
 		return nil
 	}
 
-	p.init(nil, opts, durability, statsReporter)
+	p.init(ctx, nil, opts, durability, statsReporter)
 	if iter == nil {
 		// TODO(sumeer): fix after bumping to latest Pebble.
-		innerIter, err := handle.NewIter(&p.options)
+		innerIter, err := handle.NewIterWithContext(ctx, &p.options)
 		if err != nil {
 			return err
 		}
@@ -222,8 +222,10 @@ func (p *pebbleIterator) initReuseOrCreate(
 }
 
 // setOptions updates the options for a pebbleIterator. If p.iter is non-nil, it
-// updates the options on the existing iterator too.
-func (p *pebbleIterator) setOptions(opts IterOptions, durability DurabilityRequirement) {
+// updates the options on the existing iterator too, and set the context.
+func (p *pebbleIterator) setOptions(
+	ctx context.Context, opts IterOptions, durability DurabilityRequirement,
+) {
 	if !opts.Prefix && len(opts.UpperBound) == 0 && len(opts.LowerBound) == 0 {
 		panic("iterator must set prefix or upper bound or lower bound")
 	}
@@ -336,6 +338,7 @@ func (p *pebbleIterator) setOptions(opts IterOptions, durability DurabilityRequi
 	// Set the new iterator options. We unconditionally do so, since Pebble will
 	// optimize noop changes as needed, and it may affect batch write visibility.
 	if p.iter != nil {
+		p.iter.SetContext(ctx)
 		p.iter.SetOptions(&p.options)
 	}
 }

--- a/pkg/storage/pebble_iterator_test.go
+++ b/pkg/storage/pebble_iterator_test.go
@@ -165,7 +165,7 @@ func TestPebbleIterator_SkipPointIfOutsideTimeBounds(t *testing.T) {
 		max, err := hlc.ParseTimestamp(maxStr)
 		require.NoError(t, err)
 
-		iter.setOptions(IterOptions{
+		iter.setOptions(context.Background(), IterOptions{
 			LowerBound:   []byte{0x00}, // so setOptions doesn't complain
 			MinTimestamp: min,
 			MaxTimestamp: max,

